Correctly serialize/deserialize tags

FiniteReality opened the pull request:
Fixes #8949 

This adds the list of removed entries to the TagFile codec, and also shuffles the construction of TagFile to be done as a build step in TagBuilder, rather than doing it lazily. TagLoader has also been augmented to respect the removed entries, by adding a new forge hook.

Both of the original hooks have been marked as deprecated, as they are no longer used due to the serialization going through a Codec instead.

If desired, I can drop the commit with the test data generators to make this a smaller PR.

FiniteReality committed:
Correctly serialize tags using TagFile

FiniteReality committed:
Run test data generators

FiniteReality commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -33,17 +33,25 @@
+-         List<TagEntry> list1 = list.stream().filter((p_236444_) -> {
++         // FORGE: Build TagFile directly 
++         TagFile file = p_236450_.buildTagFile();
++         List<TagEntry> list1 = Stream.concat(file.f_215959_().stream(), file.removeEntries().stream()).filter((p_236444_) -> {
```
I believe the previous behaviour was to treat removal entries as required, so that's what I did here.

FiniteReality replied:
Update after doing some searching with GitHub's awful repo search: it does not seem that remove()-d tags were treated as required originally - should I change this back to bring that in line?

sciwhiz12 replied:
My intuitive understanding of how tag element removal should work is that they won't error if the element they're trying to remove doesn't exist; it should just be silently ignored.

FiniteReality replied:
To clarify, the way the system is implemented currently, any tags in the remove list are implicitly treated as optional when deserializing - it is only when building the list that they are treated as required, and only if they are set to required. You can call `TagBuilder.remove(TagEntry.optionalElement(resourceLocation))` to specify that a removed tag isn't required, and it will pass the validation successfully.

FiniteReality replied:
Update on this: entries listed in `remove` are now treated the same way as entries listed in `values`, meaning the vanilla way of marking a tag as optional should be used to mark an optional entry to remove.

FiniteReality committed:
Use deprecated javadoc comments for methods

FiniteReality committed:
Respond to feedback from Forge discord

FiniteReality committed:
Compute additive/subtractive sets for intuitivity

This changes a good portion of the internals of TagLoader to compute the
entire closure of additive/subtractive tag entries, in order to return
the "most intuitive" thing.

As implemented, any entry in the 'remove' set takes precedence, even if
it is added back by another mod/datapack/etc. As an example:
```
// a/data/minecraft/tags/blocks/logs.json
{ "values": [], "replace": false, "remove": ["minecraft:oak_log"] }
// b/data/minecraft/tags/blocks/logs.json
{ "values": ["minecraft:oak_log"] }
```
The result here, regardless of load order, is that the `minecraft:logs`
tag does *not* include `minecraft:oak_log`.

FiniteReality committed:
Remove unnecessary forge hook and fix javadoc

Matyrobbrt commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -33,17 +33,25 @@
+       this.f_126543_.forEach((p_236449_, p_236450_) -> {
+-         List<TagEntry> list = p_236450_.m_215904_();
+-         List<TagEntry> list1 = list.stream().filter((p_236444_) -> {
++         // FORGE: Build TagFile directly 
```
Unnecessary comment.

Matyrobbrt commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -33,17 +33,25 @@
++         List<TagEntry> list1 = Stream.concat(file.f_215959_().stream(), file.removeEntries().stream()).filter((p_236444_) -> {
              return !p_236444_.m_215940_(this.f_126540_::m_7804_, this.f_126543_::containsKey);
 -         }).toList();
 +         }).filter(this::missing).collect(Collectors.toList()); // Forge: Add validation via existing resources
```
Why not `Stream#toList`?

FiniteReality replied:
This was already in the patch, so I left it as is to make the diff smaller. I can switch to toList if that's preferred.

Matyrobbrt commented on patches/minecraft/net/minecraft/data/tags/TagsProvider.java.patch:
```diff
@@ -33,17 +33,25 @@
-             JsonElement jsonelement = TagFile.f_215958_.encodeStart(JsonOps.INSTANCE, new TagFile(list, false)).getOrThrow(false, f_126541_::error);
+-            JsonElement jsonelement = TagFile.f_215958_.encodeStart(JsonOps.INSTANCE, new TagFile(list, false)).getOrThrow(false, f_126541_::error);
 -            Path path = this.f_236439_.m_236048_(p_236449_);
++            // FORGE: Use built TagFile rather than creating a new one
```
Unnecessary comment

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagFile.java.patch:
```diff
@@ -0,0 +1,22 @@
+ import java.util.List;
+ 
+-public record TagFile(List<TagEntry> f_215959_, boolean f_215960_) {
++public record TagFile(List<TagEntry> f_215959_, boolean f_215960_, List<TagEntry> removeEntries) {
```
Please add an `EXC` entry for this overload.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagFile.java.patch:
```diff
@@ -0,0 +1,22 @@
++
+    public static final Codec<TagFile> f_215958_ = RecordCodecBuilder.create((p_215967_) -> {
+-      return p_215967_.group(TagEntry.f_215911_.listOf().fieldOf("values").forGetter(TagFile::f_215959_), Codec.BOOL.optionalFieldOf("replace", Boolean.valueOf(false)).forGetter(TagFile::f_215960_)).apply(p_215967_, TagFile::new);
++      // FORGE: Serialize removeEntries in TagFile codec
```
Unnecessary comment.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
++                     // FORGE: Specify this tag is being added
++                     list.add(new TagLoader.EntryWithSource(p_215997_, s1, true));
++                  });
++                  // FORGE: Add removed tags
```
Unnecessary tag.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
++   }
++
++   // FORGE: Build both the additive and subtractive sets for the given tags
++   private Either<Collection<TagLoader.EntryWithSource>, AdditiveSubtractiveSets<T>> buildSets(TagEntry.Lookup<T> lookup, List<TagLoader.EntryWithSource> entries) {
```
This could have an `EXC` entry added.

FiniteReality replied:
What should the EXC entry be in this case?

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
+ 
+-      for(TagLoader.EntryWithSource tagloader$entrywithsource : p_215980_) {
+-         if (!tagloader$entrywithsource.f_216042_().m_215927_(p_215979_, builder::add)) {
++      for(TagLoader.EntryWithSource tagloader$entrywithsource : entries) {
```
Space after the keyword.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
+-            this.m_215978_(lookup, p_215985_).ifLeft((p_215977_) -> {
 -               f_13445_.error("Couldn't load tag {} as it is missing following references: {}", p_215984_, p_215977_.stream().map(Objects::toString).collect(Collectors.joining(", ")));
++            this.buildSets(lookup, p_215985_).ifLeft((p_215977_) -> {
++               // FORGE: Log the references on separate lines for easier visibility
```
Unnecessary comment.

FiniteReality replied:
I was asked to add this by @sciwhiz12

sciwhiz12 replied:
Can confirm, this one should be left in.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
+          public Collection<T> m_214048_(ResourceLocation p_216041_) {
+-            return map.get(p_216041_);
++            // FORGE: Only return the additive sets
++            if (map.get(p_216041_) == null) { return null; }
```
Brackets can be removed

FiniteReality replied:
I could have sworn I did this, but I guess not...

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
+ 
+-   public static record EntryWithSource(TagEntry f_216042_, String f_216043_) {
++   // FORGE: Store whether this entry is in the additive or subtractive list
++   public static record EntryWithSource(TagEntry f_216042_, String f_216043_, boolean additive) {
```
Please add an `EXC` entry for the overload.

Matyrobbrt commented on patches/minecraft/net/minecraft/tags/TagLoader.java.patch:
```diff
@@ -1,11 +1,108 @@
+       }
+    }
++
++   // FORGE: Store the additive/subtractive sets in a wrapper for convenience
```
Unnecessary comment

Matyrobbrt reviewed (changes requested):
Just a quick look, but most of the `// FORGE: ` markers are not necessary.

FiniteReality committed:
Add EXC entries and fix code formatting

FiniteReality committed:
Remove unnecessary forge comments

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=9053) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=9053) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MinecraftForge/MinecraftForge?pullRequest=9053) it.</sub>
